### ISSUE TITLE
Windows: Unable to start multipassd service

### ISSUE THREAD

**Issue by Geogboe** [canonical/multipass#573]
First of all, thank you devs for bring Windows support to this exciting looking project! I might just be doing something wrong here. Details below.

### Issue
When attempting to run `multipass launch` I get the following error: 
```
launch failed: cannot connect to the multipass socket
Please ensure multipassd is running and 'localhost:50051' is accessible
```

When checking the `multipassd` service I see that is is stopped. When attempting to start the service it immediately stops after starting. 

When running `multipassd` directly I get the error below:
```
E0124 09: 52: 38.349000000 14060 server_secure_chttp2.cc: 84
] {
"created": "@1548341558.349000000",
"description": "No address added out of total 2 resolved",
"file":"..\3rd-party\grpc\src\core\ext\transport\chttp2\server\chttp2_server.cc",
"file_line": 307,
"referenced_errors": [
    {
        "created": "@1548341558.349000000",
        "description": "Failed to add port to server",
        "file":"..\3rd-party\grpc\src\core\lib\iomgr\tcp_server_windows.cc",
        "file_line": 509,
        "referenced_errors": [
            {
                "created": "@1548341558.349000000",
                "description": "OS Error",
                "file":"..\3rd-party\grpc\src\core\lib\iomgr\tcp_server_windows.cc",
                "file_line": 201,
                "os_error": "An attempt was made to access a socket in a way forbidden by its access permissions.\r\n",
                "syscall": "bind",
                "wsa_error": 10013
            }
        ]
    },
    {
        "created": "@1548341558.349000000",
        "description": "Failed to add port to server",
        "file":"..\3rd-party\grpc\src\core\lib\iomgr\tcp_server_windows.cc",
        "file_line": 509,
        "referenced_errors": [
            {
                "created": "@1548341558.349000000",
                "description": "OS Error",
                "file":"..\3rd-party\grpc\src\core\lib\iomgr\tcp_server_windows.cc",
                "file_line": 201,
                "os_error": "An attempt was made to access a socket in a way forbidden by its access permissions.\r\n",
                "syscall": "bind",
                "wsa_error": 10013
            }
        ]
    }
]
}
error: Failed to start multipass gRPC service at localhost: 50051
```

I've tried uninstalling and reinstalling with reboots but issue persists. 

Using a lower port seems to start the daemon ( or at least shows no error ) but I don't know how to tell the client to connect to it on that port:
`multipassd.exe --address localhost:5001 -V debug`

### Background info

```
multipass version
multipass  v0.5-full
```

- Windows 10 Enterprise Version 1809 Build 10.0.17763
- Connection profiles for all my NICs are set to private. 
- HyperV is installed and was present before installing multipass
- All available Windows updates installed. List below:

```
10 Hotfix(s) Installed.
[01]: KB2693643
[02]: KB4481031
[03]: KB4465477
[04]: KB4465664
[05]: KB4470502
[06]: KB4470788
[07]: KB4471331
[08]: KB4480056
[09]: KB4480979
[10]: KB4480116
```

- Hyper-V is currently working normally
- Docker desktop and VirtualBox are installed as well though not in use.
- I'm connected to a domain and running as a domain user
- Nothing in Hyper-V seems to be built

Any ideas? Thanks!

**Comment by townsend2010** [comment]
Hi @Geogboe,

Thanks for the nice words:)

My first notion is that some other process on your machine is listening on port 50051.

From an elevated command prompt (either cmd.exe or Powershell will suffice), do:
`C:\> netstat -a -b -n -o`
and look for anything listening on port 50051.

And yes, allowing the client to connect on different port is an oversight and needs to be fixed.  I will enter a separate bug report for that.

**Comment by Saviq** [comment]
Hi @Geogboe, this usually happens when the network you were connected to when installing multipass was marked Public (aka untrusted). We're missing a step (#574) in the installer to convince Windows it's OK for multipass to listen in that case.

You're connected to a domain, which may mean your situation is also slightly different, but can you verify that the network you're connected to is marked Private (aka trusted), reboot and reinstall multipass?

**Comment by townsend2010** [comment]
Regarding the client's port use, we do have a `MULTIPASS_SERVER_ADDRESS` env var that can be set and the client will look for that when it starts.

**Comment by Geogboe** [comment]
Thanks guys! And super fast responses! 

I'm going to do a deep dive into checking ports and doing the reinstall stuff that @Saviq suggested. I just wanted to quickly state that using an alternative port and the MULTIPASS_SERVER_ADDRESS env variable seems to work. Thanks @townsend2010 !

`multipassd.exe -V info --address localhost:5001`
```cmd
λ $env:MULTIPASS_SERVER_ADDRESS = 'localhost:5001'
λ  multipass.exe find
multipass launch ΓÇª   Starts an instance of           Image version
------------------------------------------------------------------
16.04                Ubuntu 16.04 LTS                20190122.1
   (or: x, xenial)
18.04                Ubuntu 18.04 LTS                20190122
   (or: b, bionic, default, lts, ubuntu)
```

**Comment by mdlarussa** [comment]
Hi, I'm having the same issue... Wifi marked as private network, port not in use.  reinstalled twice.

I'm starting it at a command prompt with:
 "D:\Program Files\Multipass\bin\multipassd.exe" --address localhost:5001

then setting the env var in powershell, before executing multipass.exe
PS C:\WINDOWS\system32> $env:MULTIPASS_SERVER_ADDRESS=5001
PS C:\WINDOWS\system32> multipass launch
PS C:\WINDOWS\system32> multipass find
PS C:\WINDOWS\system32> multipass list
PS C:\WINDOWS\system32> multipass info

nothing happens.

**Comment by townsend2010** [comment]
@mdlarussa,

Also, I take it you were having the same issue as posted in here when using the default port?

**Comment by mdlarussa** [comment]
Bingo.  

PS C:\WINDOWS\system32> multipass launch
Retrieving image: 

Thank you.

**Comment by townsend2010** [comment]
@mdlarussa,

Ok, and you checked for any processes listening on port 50051 and you didn't see any?

**Comment by mdlarussa** [comment]
I tried to fake out the service by modifying the startup command, but I couldn't get the /svc flag to fire and return to the cmd line when the port was specified.

correct, nothing running on 50051.

**Comment by townsend2010** [comment]
@mdlarussa,

Ok, thanks for the additional info.  Need to dig further as to why port 50051 is causing us permissions headaches.

**Comment by mdlarussa** [comment]
How can I redownload the image?  looks like a network hiccup or something prevented a full download.  Now I get:
PS C:\WINDOWS\system32> multipass launch
launch failed: Network timeout

**Comment by mdlarussa** [comment]
NM - figured it out...
PS C:\WINDOWS\system32> multipass purge
PS C:\WINDOWS\system32> multipass list
No instances found.
PS C:\WINDOWS\system32> multipass launch
Retrieving image: 28%

**Comment by Geogboe** [comment]
So it seems to be working for me now on the default port so let me break down what I did. 

### Troubleshooting
per @townsend2010 I ran some netstat testing:
Strangely netstat showed nothing else listening on 50051. 
`netstat -a -b -n -o | Select-String 50051`
`netstat -a -b -n -o | findstr /im 50051`
( also tried manually looking through the whole list )


### Manual listener
To see if this was red herring or something I tried to manually set up temporary listener on 50051 using the `System.Net.Sockets.TcpListener`  class in PowerShell. It threw a similar permissions error! So I did some experimenting. I found that I couldn't set up a listener between about 50000-50389, yet nothing in netstat was using ANY of those ports...very strange. ( This behavior mirrors multipassd.exe behavior )

### Possible solution
I ended up reinstalled multipass over to of itself, removed my AV, restarted and now it's working. It could have been the additional restart, the reinstall, or the AV removal--I'm not sure. But for reference my AV was Crowdstrike ( enterprise product ). I'm planning on reinstalling it and testing again. 

### My final thoughts
Maybe multipassd.exe actually DID set up the listener on 50051 the first time around. But for some reason the service didn't register correctly. Every attempt after that then threw an error because that port was already registered to the phantom service....?

I'm going to reinstall my AV and keep testing but I'd say that this seems to be resolved for me. Thank you all!

**Comment by townsend2010** [comment]
@Geogboe,

Thanks for the detailed information.  That is quite helpful.  Sounds like *something* wants to block ports at around 50000-50389 which is affecting multipassd.  I'm doing more research...

**Comment by townsend2010** [comment]
Very interesting: https://github.com/docker/for-win/issues/3171 and https://github.com/googlevr/gvr-unity-sdk/issues/1002.

**Comment by Geogboe** [comment]
@townsend2010 great find! docker/for-win#3171 especially seems to really replicate my issue. 

There's a lot of similarities:
- On Windows 10 1809
- Just upgraded to Docker for Windows 2.0.0.0.
- Running `netsh interface ipv4 show excludedportrange protocol=tcp` also shows a bunch of big excluded port ranges. 
- Running netstat -ano shows nothing using 50051.

In my scenario I didn't have docker running at the time I was troubleshooting multipass. I HAD however started and stopped the docker daemon during that session so I'm wondering if that's when it set those port exclusions. And by rebooting those exclusions got cleared, or incremented allowing multipass to work...?

Also, I reinstalled my AV and things are still working so I guess AV is ruled out.

**Comment by townsend2010** [comment]
So after a bunch of research on this, it's Hyper-V that is reserve the port, amongst many others.  For the time being, the workaround is as follows:

1. Disable Hyper-V and restart your computer.
2. Open a command prompt as administrator.
3. Run the following command:
`netsh int ip add excludedportrange protocol=tcp startport=50051 numberofports=1`
4. Re-enable Hyper-V and restart your computer.

I'm looking into if there is a better port for `multipassd` to listen on by default, but I have found that the reserved port ranges that Hyper-V reserves seems non-deterministic as there are many other instances of users having this work without issues (myself included).

**Comment by Saviq** [comment]
Hi all, we've not seen this issue in quite a while, please report back if you still encounter it.

We suspect newer Windows updates made this go away.

**Comment by reclaimed** [comment]
A pretty strange problem here:
1.  Multipass doesn't recognize the OS:
```
PS> multipass launch
launch failed: Multipass support for Hyper-V requires Windows 10
```
2. At some point, it reported there's no Hyper-V installed
```
(can't replicate the issue but trust me :) )
```
I see this message since the first install of version 0.9. I did try to reboot the PC, as well as disable and enable Hyper-V.

However, it works well if I run the second instance of multipassd.exe on port 5001:
```
Elevated PS>  multipassd.exe -V debug --logger stderr --address localhost:5001
...
PS> $env:MULTIPASS_SERVER_ADDRESS = 'localhost:5001'
PS> multipass list
No instances found.
PS> multipass launch
Launched: protective-bobolink
PS> multipass info --all
Name:           protective-bobolink
State:          Running
IPv4:           192.168.43.215
Release:        Ubuntu 18.04.3 LTS
Image hash:     6afb97af96b6 (Ubuntu 18.04 LTS)
Load:           0.08 0.18 0.08
Disk usage:     999.5M out of 4.7G
Memory usage:   570.0M out of 921.6M
PS> multipass delete protective-bobolink
PS> multipass purge
PS> multipass list
No instances found.
```

System info:
```
PS> Get-ComputerInfo  | select windowsproductname, windowsversion, HyperVisorPresent
WindowsProductName : Windows 10 Pro
WindowsVersion     : 1909
HyperVisorPresent  : True

PS> multipass version
multipass  1.0.0+win
multipassd 1.0.0+win

PS> Get-Service multipass | select name, status, starttype
Name      : multipass
Status    : Running
StartType : Automatic

Service path and args: "C:\test\Multipass\bin\multipassd.exe" /svc

Netstat:
netstat -a -b -n -o | Select-String 50051
>   TCP    127.0.0.1:50051        0.0.0.0:0              LISTENING       4188
   [multipassd.exe]
>   TCP    [::1]:5791             [::1]:50051            ESTABLISHED     14552
   [multipass.gui.exe]
>   TCP    [::1]:50051            [::]:0                 LISTENING       4188
   [multipassd.exe]
>   TCP    [::1]:50051            [::1]:5791             ESTABLISHED     4188
   [multipassd.exe]
```

**Comment by richtera** [comment]
I don't see the multipass service automatically starting recently. I can manually start it up and then things work, but it's set to Automatic and still doesn't start up.

**Comment by Saviq** [comment]
@richtera could you please check the Event Viewer for any signs why it failed to start at first? And file a new issue please?

**Comment by richtera** [comment]
@Saviq I looked at the EventViewer for multipass but I am not seeing any error events.

**Comment by c-goes** [comment]
Found another workaround:

1. repeat reboots until `netsh interface ipv4 show excludedportrange protocol=tcp` doesn't show the port 50051 in a range (these ports change with every reboot with my Windows, unclear why). this step probably takes 50-100 reboots
2. disable Hyper-V, skip reboot
3. stop Multipass service if running
4. run `netsh int ip add excludedportrange protocol=tcp startport=50051 numberofports=1`
5. enable Hyper-V

now the ports look like this

```
netsh interface ipv4 show excludedportrange protocol=tcp
Portausschlussbereiche für das Protokoll "tcp"

Startport      Endport
----------    --------
     49709       49808
     49809       49908
     49909       50008
     50051       50051     *
     50209       50308
     50309       50408
     50424       50523
     50620       50719
     50871       50970
     50971       51070
     51071       51170
     51171       51270
     51271       51370
     51471       51570
     51571       51670
     52250       52349
     52350       52449
     54032       54131
     54132       54231

* - Verwaltete Portausschlüsse.
```

**Comment by valentinmk** [comment]
This is really wierd.
I've faced same problem.
```
launch failed: cannot connect to the multipass socket
Please ensure multipassd is running and 'localhost:50051' is accessible
```
1. Then try to add in MULTIPASS_SERVER_ADDRESS=localhost:5001 as system variable (from System Information). Multipass running, but no result, microk8s tryed to connect 50051 port during instalation
2. Uninstall multipass (installed during microk8s installation)
3. Manually install multipass only (version 1.3.0). Multipass started and run correctly.
4. Install microk8s. Installation run successfully.
🤷‍♂️

**Comment by richtera** [comment]
I am also seeing the problem with the last version. I did notice though that I can manually go to the services app and startup the multipass service which was not running even thought it's selected as "Automatic" startup. I am guessing that during the initial boot it crashes or something and then it ends up not running. The only error I see for "Multipass" in the event log is "[VMImageHost] illegal value" but that happened yesterday, so I don't think it's related.

**Comment by DanteDT** [comment]
I've tried troubleshooting mentioned above, including re-installing/re-starting ... and yet ...

Hyper-V enabled on fully up-to-date Win10 Pro 2004 b.19041.508.
Nothing else listening to 50051:
```
C:\>netstat -a -b -n -o | findstr /i "50051"

```

```
C:\>multipass launch --name pop-os-lts --mem 8G --disk 50G "file://H:\_VMs_\Distributions\pop-os_20.04_amd64_intel_13.iso"
launch failed: cannot connect to the multipass socket
Please ensure multipassd is running and 'localhost:50051' is accessible
```
Running services:
```
C:\>net start | findstr /i "multi"

C:\>net start | findstr /i "hyper"
   Hyper-V Host Compute Service
```
Multipass service (set to Automatic) fails to start with error pop-up:

"Services:  The Multipass Service service on Local Computer started and then stopped. Some services stop automatically if they are not in use by other services or programs. [OK]"

If I right-click Multipass in the system tray, it report "Failure retrieving instances"

**Comment by serkanulger** [comment]
I cannot connect to multipass service and cannot list the instances on Windows 10 20H2.
I have changed ImagePath in the registry.
ImagePath in HKEY_LOCAL_MACHINE\SYSTEM\ControlSet001\Services\Multipass 
"C:\Program Files\Multipass\bin\multipassd.exe" /svc --verbosity debug --address "localhost:5001"

and set system environment variable MULTIPASS_SERVER_ADDRESS=localhost:5001
Seems that it is working..

**Comment by totoroyyb** [comment]
I also encountered this issue on Windows. No process is using port 50051, but this port indeed seems to be excluded by the system by running (just as mentioned by ppl above):

```
netsh interface ipv4 show excludedportrange protocol=tcp
```

Since this exclusion policy might be created by Hyper-V and our multipass relies on it, well, at least by default settings. Therefore, it can sometimes be a bit annoying. I am aware of the fix by changing the port to listen, but I wonder whether multipass team considers switching to another port by default?

**Comment by Saviq** [comment]
@totoroyyb the thing is that the port is _not_ excluded by default. It's unclear what creates that policy - and until we know that, any port would be equally at risk of falling into that range…

**Comment by totoroyyb** [comment]
@Saviq Thanks for the comment. In this [comment](https://github.com/canonical/multipass/issues/573#issuecomment-460359962), I saw another contributor comment something:

> So after a bunch of research on this, it's Hyper-V that is reserve the port, amongst many others. For the time being, the workaround is as follows:

Therefore, I thought it was a deterministic behavior caused by Hyper-V. If I understand you correctly, does that mean that Hyper-V sometimes just reserves a bunch of ports randomly?

**Comment by Saviq** [comment]
@totoroyyb it doesn't do it _by default_, at least. And we've not found confirmation as to why / when it does it.

All that said, selecting a port outside of the range we see Hyper-V reserving may be wise. I've filed #2525 for that, then.

**Comment by totoroyyb** [comment]
@Saviq I see. Thank you for the update and information!

**Comment by takushi1969** [comment]
check your proxy settings to prevent to use proxy when accessing localhost.

Ex.
```
PS C:\> Get-ChildItem env: |where -Property name -like "*proxy"

Name                           Value
----                           -----
ftp_proxy                      http://proxy.hogehoge.com:8080
http_proxy                     http://proxy.hogehoge.com:8080
https_proxy                    http://proxy.hogehoge.com:8080
rsync_proxy                    http://proxy.hogehoge.com:8080
PS C:\>
```
Then set no_proxy variable.
```
$env:no_proxy="localhost,127.0.0.1"
```

**Comment by richtera** [comment]
If you can I would switch to wsl2. With multipass you will have to adjust the routing. The routing is somewhere configured on the multipass ssh command, but I have not used multipass or hyperv for a while.

**Comment by summy00** [comment]
just restart multipass service, it works for me

**Comment by govindrai** [comment]
I ran into the opposite problem. :D 

It turns out: multipass and [grpc server](https://grpc.io/docs/guides/auth/#examples) recommend the same default port. 

picking a different port and moving on

**Comment by brucejo75** [comment]
I see that `50051` is an excluded port after a windows update:

```
netsh interface ipv4 show excludedportrange protocol=tcp

Protocol tcp Port Exclusion Ranges

Start Port    End Port
----------    --------
        22          22
      2222        2222     *
      5357        5357
     50000       50059     *
     56182       56281
     56282       56381
     56382       56481
     61709       61808
     61809       61908
     61909       62008

* - Administered port exclusions.
```

What is an administered port exclusion? [See this](https://stackoverflow.com/questions/58216537/what-is-administered-port-exclusions-in-windows-10)

On Windows it looks like maybe Teams needs the port:

https://answers.microsoft.com/en-us/msteams/forum/all/microsoft-teams-ipports/b21fb0e0-1226-4bcf-b527-92c810bc7a87

**Comment by fauzimh** [comment]
Hi, I'd like to share that I fixed it by killing multipassd.exe process using task manager. Afterwards, a new multipassd.exe will appear, then I simply retry the multipass launch command, and it worked.

**Comment by bushmaster570** [comment]
I upgraded multipass version to 1.12.2, (of course by uninstalling 1.11.1) on my Windows 11. After the upgrade, the multipass is unable to start. I tried to uninstall 1.12.2 and then re-install 1.11.1; but the things couldn't get back to wellness. I tried various things like disabling/rebooting/enabling/rebooting hyperV; marking my network connection as private and then re-installing but nothing has helped. I then tried to run C:\Program Files\Multipass\bin\multipassd.exe as Administrator and it seems to have started. I can see all my old instances; stop/start them. But the workaround doesn't automate the multipass start on windows boot. Can someone guide?

**Comment by bushmaster570** [comment]
> I upgraded multipass version to 1.12.2, (of course by uninstalling 1.11.1) on my Windows 11. After the upgrade, the multipass is unable to start. I tried to uninstall 1.12.2 and then re-install 1.11.1; but the things couldn't get back to wellness. I tried various things like disabling/rebooting/enabling/rebooting hyperV; marking my network connection as private and then re-installing but nothing has helped. I then tried to run C:\Program Files\Multipass\bin\multipassd.exe as Administrator and it seems to have started. I can see all my old instances; stop/start them. But the workaround doesn't automate the multipass start on windows boot. Can someone guide?

I managed to start multipass. I had previously set a non-default storage place for multipass instances. I changed it back to default. Opened power shell as administrator and ran:

PS> Stop-Service Multipass
PS> Remove-ItemProperty -Path "HKLM:System\CurrentControlSet\Control\Session Manager\Environment" -Name MULTIPASS_STORAGE

Then rebooted windows.

The multipass now seems to start properly. However, if I change the path back from default to custom:

PS> Stop-Service Multipass
PS> Set-ItemProperty -Path "HKLM:System\CurrentControlSet\Control\Session Manager\Environment" -Name MULTIPASS_STORAGE -Value "E:\Multipass"

It fails to start again.

So I rolled back the changes. Backed up old E:\Multipass and copied everything from C:\ProgramData\Multipass to E:\Multipass, changed path to custom storage. Then I copied my instances from backup to E:\Multipass\data\vault\instances. I also copied the files "E:\Multipass-backup\data\vault\multipassd-instance-image-records.json", "E:\Multipass-backup\data\multipassd-vm-instances.json" and rebooted. multipass starts well but I can't see my old instances:

>multipass list
No instances found.

The file "E:\Multipass\data\multipassd-vm-instances.json" that I copied from backup looks empty again after multipass restart.

Now the problem is how to get my instance back and running in multipass.

The multipass upgrade is definitely not as simple as should be :(.
